packages/utils/src/normalize.ts

```diff
@@ -51,170 +53,173 @@ export function normalizeToSize<T>(
 }
 
 /**
- * Walks an object to perform a normalization on it
+ * Visits a node to perform a normalization on it
  *
- * @param key of object that's walked in current iteration
- * @param value object to be walked
+ * @param key The key corresponding to the given node
+ * @param value The node to be visited
  * @param depth Optional number indicating how deep should walking be performed
- * @param maxProperties Optional maximum  number of properties/elements included in any single object/array
+ * @param maxProperties Optional maximum number of properties/elements included in any single object/array
  * @param memo Optional Memo class handling decycling
  */
-export function walk(
+export function visit(
   key: string,
-  value: UnknownMaybeWithToJson,
+  value: unknown,
   depth: number = +Infinity,
   maxProperties: number = +Infinity,
   memo: MemoFunc = memoBuilder(),
-): unknown {
+): Primitive | { [key: string]: unknown } {
   const [memoize, unmemoize] = memo;
 
-  // If we reach the maximum depth, serialize whatever is left
-  if (depth === 0) {
-    return serializeValue(value);
+  // if the value has a `toJSON` method, bail and let it do the work
+  const valueWithToJSON = value as unknown & { toJSON?: () => string };
+  if (valueWithToJSON && typeof valueWithToJSON.toJSON === 'function') {
+    try {
+      return valueWithToJSON.toJSON();
+    } catch (err) {
+      return `**non-serializable** (${err})`;
```

Comment:
> * Instead of aborting here when `toJSON()` throws, we could continue and try to serialize the object ourselves.

Good idea.

> * In `normalize`, we return an object `{ ERROR: ... }` in case of an error, here we just return a string. Should we align those return values?

If we get rid of the `**non-serializable**` instance above, the only other one is inside of `stringifyValue`, and I'd rather keep that one a string because it's (literally) the name of the game at that point in the code.

Reply:
Seems fair!